### DISCUSSION TITLE
Add a version of mul/div that truncates rather than rounds

### DISCUSSION THREAD

**Discussion by JasoonS** [PaulRBerg/prb-math#126] [Ideas]
In our code we require that multiplication rounds down always.

Is there a way to use this library with that in mind? Would there be plans to do that?

Thanks

As an example in the following code:
```
      console2.log("prbMath", uint256(630387309078063512139).mul(1030548320346703977));
      console2.log("standard", (uint256(630387309078063512139) * 1030548320346703977) / 1e18);
```

Prints
```
  prb-math 649644582538276888396
  standard 649644582538276888395
```
The actual result is 649.644582538276888395838879597179076803  (with 18 decimal precision) so it is shown it is rounding up (not taking the floor), but I see the docs for `mulDivFixedPoint` say:
```
    /// @notice Calculates floor(x*y÷1e18) with full precision.
```

**Comment by JasoonS** [comment]
I did some experimentation with the open-zeppelin maths library too:

```
    console2.log("actual   ", "649644582538276888395.838879597179076803");
    console2.log("prb-math ", PRBMathUD60x18.mul(630387309078063512139, 1030548320346703977));
    console2.log("standard ", (uint256(630387309078063512139) * 1030548320346703977) / 1e18);
    console2.log("OZ - std ", Math.mulDiv(630387309078063512139, 1030548320346703977, 1e18));
    console2.log("OZ - up  ", Math.mulDiv(630387309078063512139, 1030548320346703977, 1e18, Math.Rounding.Up));
    console2.log("OZ - down", Math.mulDiv(630387309078063512139, 1030548320346703977, 1e18, Math.Rounding.Down));
    console2.log("OZ - zers", Math.mulDiv(630387309078063512139, 1030548320346703977, 1e18, Math.Rounding.Zero));
```

This prints:
```
  actual    649644582538276888395.838879597179076803
  prb-math  649644582538276888396
  standard  649644582538276888395
  OZ - std  649644582538276888395
  OZ - up   649644582538276888396
  OZ - down 649644582538276888395
  OZ - zers 649644582538276888395
```

**Comment by PaulRBerg** [comment]
Hey @JasoonS, thanks very much for opening a discussion about this! Feedback is always welcome, since it helps me improve the usefulness of the library.

I will be upfront, I don't have any plans for adding support different types of rounding modes in V3 (the next version of PRBMath, which I will ship soon-ish). As explained in the [NatSpec](https://github.com/paulrberg/prb-math/blob/88d0815baef78c0699fec1ff10b34a35903e110f/contracts/PRBMath.sol#L480-L497) of the `mulDivFixedPoint` function, I have relied upon the logic of [listing 6 from this article on ACCU](https://accu.org/journals/overload/18/100/harris_1717/) when I wrote the multiplication function. That is so values like 6.6e-19 are rounded up to 1e-18 rather than being truncated to 0.

However, if I keep receiving requests for rounding modes, I may consider implementing it in PRBMath V4.

I'm curious. May I ask you why do you "require" that multiplication rounds down, always?

**Comment by JasoonS** [comment]
Thanks @paulrberg .

> I'm curious. May I ask you why do you "require" that multiplication rounds down, always?

I was looking at using this for https://float.capital v2. We do asynchronous maths on a per user basis and on a batched basis with multiplication division. If the sum of all the per user operations is greater than the operation on the whole batch we are at risk of overflow.

More specifically, we calculate how much erc20 token to mint for all the users in the batch with 1 multiplication on the batch. But then each user individually claims those tokens in their own time later. If the sum of all the user claims is greater than the batch the contract breaks. Rounding down makes this impossible (although some dust tokens will accumulate over time).

Really awesome stuff on the library - very useful tool for the box :raised_hands: Thanks!

**Reply by PaulRBerg** [reply]
Thanks a lot for sharing your feedback, @JasoonS. This is an interesting use case. I agree with you that rounding down would be better in your case.

I will consider this feature in future versions of the library. Thanks for being a user :)

**Comment by PaulRBerg** [comment]
Hi @JasoonS, I have recently released PRBMath [v3.0.0](https://github.com/paulrberg/prb-math/releases/tag/v3.0.0), and I am happy to let you know that I listened to your feedback here and changed the behavior of the `mul` function to always truncate rather than round up. Related: https://github.com/paulrberg/prb-math/discussions/128.

Does this address your need?

**Reply by JasoonS** [reply]
Hi Paul, thank you - that is great news :blush:
